Source/WebCore/platform/graphics/PlatformImage.h

```diff
@@ -68,6 +70,16 @@ class BitmapRef: public BBitmap, public RefCounted<BitmapRef>
         ~BitmapRef()
         {
         }
+
+        // TODO: Is it possible to just use the BBitmap's address in memory for
```

Comment:
I don't know what this is used for, but:

- Either the ID is valid within a single process, and in that case, you can indeed use the address of the bitmap,
- Or, the ID is used to instanciate a BitmapRef in another process, with both BitmapRefs referencing the same bitmap. In that case, this won't work.

In any case, the role of BitmapRef is very specific: to make it possible to use BBitmap as a reference-counted object in WebKit classes. So, if you have some new fields and functionalites to add, this is almost certainly not the right place to do it. Either add it in BBitmp in Haiku (if there is no option to do it in WebKit), or add it in some other class in WebKit that is meant to store some platform specific data.

Reply:
It seems to be used to see if the BBitmap used to back a graphics layer has changed. It also seems like it needs to work across processes. So it doesn't look like the memory address would work.

It seems there are three good approaches to getting a unique id:

1. Use the area_id + offset of BBitmap. This information can be obtained without modifying BBitmap. There is a *very* tiny chance of collision.
2. Wrap the class to keep the uniqueID function.
3. Expose BBitmap::_ServerToken.

Which way sounds best?

Reply:
I think area_id + offset should be fine for now. This is the info you need to recreate a valid BBitmap in the other process that will share the same memory.

Reply:
And for completeness: it seems in cairo this is implemented as a separate function: getSurfaceUniqueID so we don't need to modify the class here.

Reply:
Cairo's getSurfaceUniqueID seems to use a function Cairo implements that allows storing custom data with the Cairo surface. So, no workaround there.

IIRC, area_id (32 bit) and offset (64 bit) will need to be combined into a single 64 bit number. So there is a really tiny chance of collision if conditions are just right (likely involving areas that are larger than 4 GB). Other than that, it should work perfectly fine.

Reply:
It's unlikely that the offset will be that large and go past 4GB. The main risk of collision would be if a bitmap was freed and a new one allocated at the same address in the same area.

I think this should be good enough for now, let's get the thing working (and this change merged) and we can modify it again later if needed.

Reply:
Looks like the current implementation never offsets the bitmaps, so the area_id alone should be sufficient! https://github.com/haiku/haikuwebkit/blob/haiku-webkit2/Source/WebCore/platform/graphics/haiku/ShareableBitmapHaiku.cpp#L74

But now I'm looking into whether the area_id is actually a good identifier. A process who wants to use an area has to first clone it. But that would mean they have different area_ids. So will the area_id of a shared image actually be the same across both processes?

Reply:
Looks like I can't use area_id. I shared a BBitmap across two simple programs and they both had different area_ids. I'm now looking into base64 encoding an identifier into the name of the memory area when it is created.

Reply:
Now I realize the method I was using likely wouldn't have worked anyway. I doubt BitmapRef's unique id would have been synchronized across processes. Anyway, this is now done.